lib/Dialect/Lattigo/Transforms/AllocToInplace.cpp

```diff
@@ -3,98 +3,223 @@
 #include <utility>
 
 #include "lib/Dialect/Lattigo/IR/LattigoOps.h"
-#include "mlir/include/mlir/Analysis/Liveness.h"  // from @llvm-project
-#include "mlir/include/mlir/IR/MLIRContext.h"     // from @llvm-project
-#include "mlir/include/mlir/IR/PatternMatch.h"    // from @llvm-project
-#include "mlir/include/mlir/Support/LLVM.h"       // from @llvm-project
-#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+#include "lib/Utils/Tablegen/InplaceOpInterface.h"
+#include "mlir/include/mlir/Analysis/Liveness.h"        // from @llvm-project
+#include "mlir/include/mlir/Dialect/Func/IR/FuncOps.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/MLIRContext.h"           // from @llvm-project
+#include "mlir/include/mlir/IR/PatternMatch.h"          // from @llvm-project
+#include "mlir/include/mlir/Support/LLVM.h"             // from @llvm-project
+#include "mlir/include/mlir/Transforms/WalkPatternRewriteDriver.h"  // from @llvm-project
 
 namespace mlir {
 namespace heir {
 namespace lattigo {
 
+// There are two types of Value in the IR:
+// 1. Storage: the actual memory allocated for the value
+// 2. Referring Value: the value that refers to the storage (e.g., the
+// returned SSA of an inplace operation)
+//
+// This class is similar to Disjoint-set data structure.
+// Each Storage is the root, and all Referring Values are in its set.
+//
+// At the beginning StorageInfo should be initialized based on current relation
+// in program.
+//
+// During rewriting, when we find available Storage for an AllocOp, we replace
+// it with an InplaceOp and update the StorageInfo by merging the Storage of the
+// AllocOp to the available Storage.
+//
+// This allows a mix of AllocOp and InplaceOp in input IR for the pass.
+class StorageInfo {
+ public:
+  StorageInfo() = default;
+
+  void addStorage(Value value) {
+    if (mlir::isa<RLWECiphertextType>(value.getType())) {
+      storageToReferringValues[value] = {};
+    }
+  }
+
+  void addReferringValue(Value storage, Value value) {
+    storageToReferringValues[storage].push_back(value);
+  }
+
+ private:
+  // maintenance should be called internally
+  void removeStorage(Value value) { storageToReferringValues.erase(value); }
+
+  void mergeStorage(Value from, Value to) {
+    storageToReferringValues[to].insert(storageToReferringValues[to].end(),
```

Comment:
I'm guessing you need a reserve/resize here so you're not inserting past the bounds of the array?

Reply:
Added a reserve here

Reply:
ok that was my guess. I will be able to dig in a bit deeper internally tomorrow or later tonight.